torchgen/dest/register_dispatch_key.py

```diff
@@ -615,6 +615,8 @@ def gen_class_set_output_body(self, k: SchemaKind, maybe_create_proxy: bool) ->
   guard_.reset_device(options.device());
 }
 """
+            if self.backend_index.dispatch_key == DispatchKey.CUDA:
+                maybe_set_guard = "c10::cuda::SetTargetDevice();" + maybe_set_guard
```

Comment:
you are unconditionally initializing context on a device that is pointed to by targetDeviceIndex (not just on `options.device()`, which is what intended).